lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php

```diff
@@ -283,7 +283,7 @@ public function getListTableForeignKeysSQL($table, $database = null)
     {
         return 'SELECT quote_ident(r.conname) as conname, pg_catalog.pg_get_constraintdef(r.oid, true) as condef
                   FROM pg_catalog.pg_constraint r
-                  WHERE r.conrelid =
+                  WHERE r.conrelid IN
```

Comment:
I think the right fix would be to trigger a deprecation if `$table` doesn't contain a dot, because that's just an incorrect usage. I think @morozov is right, listing all the foreign keys of all the tables named like `$table` makes no sense. Right now, this doesn't crash when there is only one table, right?

Reply:
The main issue here is that this **is** a bug that needs to be fixed. A WHERE clause that is comparing against a sub SELECT statement must either be an IN comparison or if it's an equals (=) comparison then the SELECT statement **must** return only one row (i.e. LIMIT 1). The current code does not comply with those SQL rules, and that is what this PR is intended to address.

The theology of what the `getListTableForeignKeysSQL()` should return is a different discussion. Yes, it is related to how this fix could be implemented. And I'm willing to have that discussion. But I'm not just going arbitrarily pick a direction.

Here is my opinion - it is **not** incorrect usage to pass just the table name without the schema, as listing all foreign keys for the table that is passed is what the function is built to do. The `getListTableForeignKeysSQL()` function is used in the `AbstractSchemaManager->listTableForeignKeys()` and this is the description in the DocBlock for the purpose of this function: `Lists the foreign keys for the given table.` It might _"make no sense"_ to you, but as odd as it may be, someone might find a use case for getting a list of all foreign keys for a table name across multiple schemas. So, IMHO, this is fine as it is.

But if you feel strongly about limiting this function to require a schema in the table name (in the context of postresql), then `AbstractSchemaManager->listTableForeignKeys()` should be overridden in `PostgreSqlSchemaManager` with the check for the dot in the `$table` in there (not in the `getListTableForeignKeysSQL()` function), and the DocBlock description changed to 
`Lists the foreign keys for the given table in a given schema.`
`@param string      $table    The name of the table including schema ('some_schema.some_table').`

Reply:
> someone might find a use case for getting a list of all foreign keys for a table name across multiple schemas.

But that person cannot be using the feature at the moment, since it will inevitably crash, correct? 

That means that we could:

1. catch the current exception when it happens, and wrap it in an `\InvalidArgumentException` with a meaningful message if and only if the database name is not provided by the user. If somebody can think of a use case for this, they will see the meaningful message, and we will use `IN`, and add a test that document the use case.
2. when the exception does not happen, trigger a deprecation informing the user that they should include the database name.
3. once that is merged, we will remove the deprecation and throw an exception as soon as the database name is not provided in the next major version of this library. 

I agree with you about the override of `listTableForeignKeys` but it should only trigger a deprecation for now. The docblock could be changed as you suggested since that is not a BC-break.

WDYT?

Reply:
Just to provide clarity - you were using the term "database name" in your last comment, when I believe you meant "schema name"

After digging into this further, the `getTableWhereClause()` function is where queries using that function could end up returning multiple results from same table names across multiple schemas, via this line of code:
`$schema = "ANY(string_to_array((select replace(replace(setting,'\"\$user\"',user),' ','') from pg_catalog.pg_settings where name = 'search_path'),','))";` _(which means: any schema in the db user search path)_

And the `getTableWhereClause()` function is used in 3 functions in the `PostgreSqlPlatform` class:
`getListTableIndexesSQL()` which uses the IN comparison in the WHERE clause,
`getListTableColumnsSQL()`,
and `getListTableForeignKeysSQL()` 

So, if you're wanting to test for a dot in the `$table` parameter to trigger a depreciation message, then it should be done in the `getTableWhereClause()` function. Then in the next major version remove the fall back `$schema = 'any schema in db user search path'` and throw an exception if the `$table` doesn't include the schema.

I still say that the malformed SQL in the `getListTableForeignKeysSQL()` needs to be fixed. The `getListTableIndexesSQL()` uses the IN comparison exactly as this PR proposes. So there is precedent, and this leads me to believe that the original developer may have made a mistake in using the = comparison in the `getListTableForeignKeysSQL()` function.

Reply:
If we use `IN`, then we will start "technically correct" results. That sounds like malicious compliance to me. The fact that there's a precedent would lead me to question the code in `getListTableIndexesSQL`. You can change my mind by showing a plausible use case for getting objects for all the tables with the same name in different schemas, maybe I'm just lacking imagination.

Reply:
> You can change my mind by showing a plausible use case for getting objects for all the tables with the same name in different schemas

For me personally, I don't see a "plausible" use case for this (but someone might). If you are wanting to require the schema be given in the `$table` parameter - I believe that is that is reasonable considering that a "plausible" use case would be rare. I just don't think it's the right approach to use the error resulting from malformed SQL to check for the schema requirement - as you suggested. That's essentially using a bug to validate requirements.

> If we use IN, then we will start "technically correct" results. That sounds like malicious compliance to me. The fact that there's a precedent would lead me to question the code in getListTableIndexesSQL.

You don't have to use `IN` to have the SQL be "compliant" with the rules. You can use `LIMIT 1` in the sub select in the `WHERE = (SELECT ...)` clause. 

So, this is what I suggest, and I'll make these changes if you agree:

- Change the queries in `getListTableForeignKeysSQL` and `getListTableIndexesSQL` to use `LIMIT 1`
- Have the `getTableWhereClause` validate that `$table` is prefixed with a schema, and if it doesn't, trigger a depreciation.
- In the next major version, remove the `$schema` fall back query: `$schema = "ANY(string_to_array((select replace(replace(setting,'\"\$user\"',user),' ','') from pg_catalog.pg_settings where name = 'search_path'),','))"` in the `getTableWhereClause`

Reply:
Sounds almost perfect to me. I would remove the `LIMIT 1` too in the next major version, but maybe that was implied.

Reply:
Ok, I'll work on implementing this. Why would you want to remove `LIMIT 1` from in the next version?

Reply:
Because a crash would be an indication that something is wrong, just like it made you detect that something was wrong right now (not limiting the query to a particular schema).